### DISCUSSION TITLE
Can't Seem to Enable DB Connection Pooling

### DISCUSSION THREAD

**Discussion by abotelho-cbn** [quay/quay#3223] [Q&A] [answered]
Hi all,

I've been reading through documentation quite a bit, and I've found hints about enabling DB connection pooling.

This seems to imply setting an environment variable:
https://docs.redhat.com/en/documentation/red_hat_quay/3.5/html/manage_red_hat_quay/advanced-quay-configuration#adjust_database_connection_pooling

Which I have:
```
[LIVE] quay@registry1:~> podman exec -it quay /bin/bash
bash-4.4$ env
PYTHONUNBUFFERED=1
LC_ALL=C.UTF-8
HTTP_PROXY=http://proxy.example.com:3128
IGNORE_VALIDATION=false
WORKER_COUNT_SECSCAN=4
DB_CONNECTION_POOLING=true
LANG=C.UTF-8
TZ=UTC
WORKER_COUNT_REGISTRY=64
QUAYDIR=/quay-registry
QUAYCONF=/quay-registry/conf
PYTHONIOENCODING=UTF-8
container=oci
QUAYPATH=/quay-registry
USERS_DEBUG=0
PWD=/quay-registry
HOME=/quay-registry
https_proxy=http://proxy.example.com:3128
http_proxy=http://proxy.example.com:3128
no_proxy=localhost, 127.0.0.1
HTTPS_PROXY=http://proxy.example.com:3128
TERM=xterm
WORKER_COUNT_WEB=32
SHLVL=1
PYTHONPATH=/quay-registry
DEBUGLOG=false
QUAY_OVERRIDE_SERVICES=repomirrorworker=true
PYTHONUSERBASE=/app
PATH=/app/bin/:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
DB_ALLOW_POOLING=true
ftp_proxy=http://proxy.example.com:3128
QUAYRUN=/quay-registry/conf
_=/usr/bin/env
```

This seems to imply the same thing: https://github.com/quay/quay/blob/master/data/database.py
This too: https://docs.projectquay.io/config_quay.html

A change in 3.12.0 (I am running 3.12.2 in a Podman container) seems to imply you can also add it to config.yaml: https://github.com/quay/quay/pull/2783. This change also mentions a DB_ALLOW_POOLING parameter, which I enabled as an environment variable, and a configuration parameter.

```
DB_ALLOW_POOLING: true
DB_CONNECTION_POOLING: true
DB_CONNECTION_ARGS:
    max_connections: 500
```

But *still*, ```podman logs quay``` includes the following line occasionally:
```
gunicorn-web stdout | 2024-09-11 21:26:31,170 [313] [INFO] [data.database] Connection pooling disabled for postgresql
```

So I have *no* idea where else I am supposed to be enabling connection pooling. Does anyone have any ideas?

**Comment by ibazulic** [comment]
Health checks will not be pooled, for a very good reason: we want to know if the database is down at the moment of checks. What you want to see is whether pooling is initialized when gunicorn workers start. That is early in the boot process. You also do not need 

```
DB_ALLOW_POOLING: true
DB_CONNECTION_POOLING: true
```

since the env. variable explicitly enables pooling of connections. The `max_connections` sets the maximum number of connections per worker, not per pod. Because each worker initializes its own pool. So with your current setup, the maximum amount of connections this Quay instance could initialize is insanely large. You have 32 web workers and 64 registry workers. Add to that 16 secscan and around 35 others. That would mean roughly 150 processes, each having an internal pool of 500 connections. Which is...a lot 😄 You can limit yourself to `max_connections: 10` or even 5, that will most likely be enough for most purposes, especially considering how many workers you actually start.

**Reply by abotelho-cbn** [reply]
How do I know that that stdout output refers to healthchecks only? I guess that wasn't clear to me, and seemed like connection pooling was disabled globally.

**Reply by ibazulic** [reply]
Well, we start pooling by default, unless you specify it otherwise:

https://github.com/quay/quay/blob/0a1d97463f58be6968adcc8371e7fc32c4d84df6/data/database.py#L491-L496

So if the env. variable is not there, then pooling will be turned on. You have to explicitly tell Quay to turn it off. As for the health check, you cannot see that it is a health check just by simply looking at that one line. You need a bit more context from the logs. But as I said, if you want to be certain, look at the logs from early startup, when these workers come online. There will be a line saying connection pooling is enabled and the pool size and the connection stale time.

**Reply by ibazulic** [reply]
And this is where we read the value from the environment:

https://github.com/quay/quay/blob/0a1d97463f58be6968adcc8371e7fc32c4d84df6/data/database.py#L446-L460

So you should see that message, listed in this part of the code in the Quay logs.

**Reply by abotelho-cbn** [reply]
Ah, it seems like I must have misread documentation that said it was *disabled* by default.

I'll just remove these parameters then, it sounds like the defaults will be fine.

Cheers!

**Reply by ibazulic** [reply, marked as answer]
No worries :-) The docs you were referring to are old, so there might be errors in there. I'll take a closer look at the documentation to see if anything needs changing!

'ave fun!
